File: library/cpp/tvmauth/client/misc/roles/roles.cpp
#include "roles.h" 
 
#include <library/cpp/tvmauth/checked_service_ticket.h> 
#include <library/cpp/tvmauth/checked_user_ticket.h> 
 
namespace NTvmAuth::NRoles { 
    TRoles::TRoles(TMeta&& meta, 
                   TTvmConsumers tvm, 
                   TUserConsumers user, 
                   TRawPtr raw) 
        : Meta_(std::move(meta)) 
        , TvmIds_(std::move(tvm)) 
        , Users_(std::move(user)) 
        , Raw_(std::move(raw)) 
    { 
        Y_ENSURE(Raw_); 
    } 
 
    TConsumerRolesPtr TRoles::GetRolesForService(const TCheckedServiceTicket& t) const { 
        Y_ENSURE_EX(t, 
                    TIllegalUsage() << "Service ticket must be valid, got: " << t.GetStatus()); 
        auto it = TvmIds_.find(t.GetSrc()); 
        return it == TvmIds_.end() ? TConsumerRolesPtr() : it->second; 
    } 
 
    TConsumerRolesPtr TRoles::GetRolesForUser(const TCheckedUserTicket& t, 
                                              std::optional<TUid> selectedUid) const { 
        Y_ENSURE_EX(t, 
                    TIllegalUsage() << "User ticket must be valid, got: " << t.GetStatus()); 
        Y_ENSURE_EX(t.GetEnv() == EBlackboxEnv::ProdYateam, 
                    TIllegalUsage() << "User ticket must be from ProdYateam, got from " << t.GetEnv()); 
 
        TUid uid = t.GetDefaultUid(); 
        if (selectedUid) { 
            auto it = std::find(t.GetUids().begin(), t.GetUids().end(), *selectedUid); 
            Y_ENSURE_EX(it != t.GetUids().end(), 
                        TIllegalUsage() << "selectedUid must be in user ticket but it's not: " 
                                        << *selectedUid); 
            uid = *selectedUid; 
        } 
 
        auto it = Users_.find(uid); 
        return it == Users_.end() ? TConsumerRolesPtr() : it->second; 
    } 
 
    const TRoles::TMeta& TRoles::GetMeta() const { 
        return Meta_; 
    } 
 
    const TString& TRoles::GetRaw() const { 
        return *Raw_; 
    } 
 
    bool TRoles::CheckServiceRole(const TCheckedServiceTicket& t, 
                                  const TStringBuf roleName) const { 
        TConsumerRolesPtr c = GetRolesForService(t); 
        return c ? c->HasRole(roleName) : false; 
    } 
 
    bool TRoles::CheckUserRole(const TCheckedUserTicket& t, 
                               const TStringBuf roleName, 
                               std::optional<TUid> selectedUid) const { 
        TConsumerRolesPtr c = GetRolesForUser(t, selectedUid); 
        return c ? c->HasRole(roleName) : false; 
    } 
 
    bool TRoles::CheckServiceRoleForExactEntity(const TCheckedServiceTicket& t, 
                                                const TStringBuf roleName, 
                                                const TEntity& exactEntity) const { 
        TConsumerRolesPtr c = GetRolesForService(t); 
        return c ? c->CheckRoleForExactEntity(roleName, exactEntity) : false; 
    } 
 
    bool TRoles::CheckUserRoleForExactEntity(const TCheckedUserTicket& t, 
                                             const TStringBuf roleName, 
                                             const TEntity& exactEntity, 
                                             std::optional<TUid> selectedUid) const { 
        TConsumerRolesPtr c = GetRolesForUser(t, selectedUid); 
        return c ? c->CheckRoleForExactEntity(roleName, exactEntity) : false; 
    } 
 
    TConsumerRoles::TConsumerRoles(THashMap<TString, TEntitiesPtr> roles) 
        : Roles_(std::move(roles)) 
    { 
    } 
 
    bool TConsumerRoles::CheckRoleForExactEntity(const TStringBuf roleName, 
                                                 const TEntity& exactEntity) const { 
        auto it = Roles_.find(roleName); 
        if (it == Roles_.end()) { 
            return false; 
        } 
 
        return it->second->Contains(exactEntity); 
    } 
 
    TEntities::TEntities(TEntitiesIndex idx) 
        : Idx_(std::move(idx)) 
    { 
    } 
} 

File: library/java/tvmauth/src/main/java/ru/yandex/passport/tvmauth/roles/Roles.java
package ru.yandex.passport.tvmauth.roles;

import java.util.Arrays;
import java.util.Map;

import javax.annotation.Nonnull;

import ru.yandex.passport.tvmauth.BlackboxEnv;
import ru.yandex.passport.tvmauth.CheckedServiceTicket;
import ru.yandex.passport.tvmauth.CheckedUserTicket;
import ru.yandex.passport.tvmauth.exception.NotAllowedException;

public class Roles {
    private Meta meta;
    private Map<Integer, ConsumerRoles> tvm;
    private Map<Long, ConsumerRoles> user;
    private String raw;

    public Roles(
        @Nonnull Meta meta,
        @Nonnull Map<Integer, ConsumerRoles> tvm,
        @Nonnull Map<Long, ConsumerRoles> user,
        @Nonnull String raw) {
        this.meta = meta;
        this.tvm = tvm;
        this.user = user;
        this.raw = raw;
    }

    @Nonnull
    public Meta getMeta() {
        return this.meta;
    }

    @Nonnull
    public String getRaw() {
        return this.raw;
    }

    public ConsumerRoles getRolesForService(@Nonnull CheckedServiceTicket checked) {
        if (!checked.booleanValue()) {
            throw new NotAllowedException("ServiceTicket is not valid");
        }

        return this.tvm.get(checked.getSrc());
    }

    public ConsumerRoles getRolesForUser(@Nonnull CheckedUserTicket checked) {
        commonUserChecks(checked);

        return this.user.get(checked.getDefaultUid());
    }

    public ConsumerRoles getRolesForUser(@Nonnull CheckedUserTicket checked, long selectedUid) {
        commonUserChecks(checked);
        if (!Arrays.stream(checked.getUids()).anyMatch(x -> x == selectedUid)) {
            throw new NotAllowedException("selectedUid must be in user ticket but it's not: " + selectedUid);
        }

        return this.user.get(selectedUid);
    }

    private void commonUserChecks(@Nonnull CheckedUserTicket checked) {
        if (!checked.booleanValue()) {
            throw new NotAllowedException("UserTicket is not valid");
        }
        if (checked.getEnv() != BlackboxEnv.PROD_YATEAM) {
            throw new NotAllowedException("User ticket must be from ProdYateam, got from " + checked.getEnv());
        }
    }

    // some shortcuts

    public boolean checkServiceRole(@Nonnull CheckedServiceTicket checked, @Nonnull String role) {
        ConsumerRoles c = getRolesForService(checked);
        if (c == null) {
            return false;
        }

        return c.hasRole(role);
    }

    public boolean checkUserRole(@Nonnull CheckedUserTicket checked, @Nonnull String role) {
        ConsumerRoles c = getRolesForUser(checked);
        if (c == null) {
            return false;
        }

        return c.hasRole(role);
    }

    public boolean checkUserRole(@Nonnull CheckedUserTicket checked, @Nonnull String role, long selectedUid) {
        ConsumerRoles c = getRolesForUser(checked, selectedUid);
        if (c == null) {
            return false;
        }

        return c.hasRole(role);
    }
}
